pkg/pgmodel/nan_test.go

```diff
@@ -0,0 +1,175 @@
+	res := resp.GetResults()
+	if len(res) != 1 {
+		t.Fatal("Expect one result")
+	}
+	ts := res[0].GetTimeseries()
+	if len(ts) != 1 {
+		t.Fatal("Expect one timeseries")
+	}
+	samples := ts[0].GetSamples()
+	if len(samples) != 1 {
+		t.Fatal("Expect one sample")
+	}
+	return samples[0].GetValue()
+}
+
+func getBooleanSQLResult(t *testing.T, db *pgxpool.Pool, sql string) bool {
+	var res *bool = nil
+	err := db.QueryRow(context.Background(), sql).Scan(&res)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if res == nil {
+		t.Fatalf("NULL found")
+	}
+	return *res
+}
+
+func TestSQLStaleNaN(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping integration test")
+	}
+
+	withDB(t, *database, func(db *pgxpool.Pool, t *testing.T) {
+		metricName := "StaleMetric"
+		metrics := []prompb.TimeSeries{
+			{
+				Labels: []prompb.Label{
+					{Name: metricNameLabelName, Value: metricName},
+					{Name: "foo", Value: "bar"},
+					{Name: "common", Value: "tag"},
+					{Name: "empty", Value: ""},
+				},
+				Samples: []prompb.Sample{
+					{Timestamp: 10, Value: 0.1},
+					{Timestamp: 20, Value: math.Float64frombits(value.StaleNaN)},
+					{Timestamp: 30, Value: math.NaN()},
+					{Timestamp: 40, Value: 0.4},
+				},
+			},
+		}
+
+		ingestor := NewPgxIngestor(db)
+		defer ingestor.Close()
+		_, err := ingestor.Ingest(metrics)
+
+		if err != nil {
+			t.Fatalf("unexpected error while ingesting test dataset: %s", err)
+		}
+
+		query := []struct {
+			rrq        prompb.ReadRequest
+			isNaN      bool
+			isStaleNaN bool
+		}{
+			{
+				isStaleNaN: true,
+				rrq: prompb.ReadRequest{
+					Queries: []*prompb.Query{
+						{
+							Matchers: []*prompb.LabelMatcher{
+								{
+									Type:  prompb.LabelMatcher_EQ,
+									Name:  metricNameLabelName,
+									Value: metricName,
+								},
+							},
+							StartTimestampMs: 19,
+							EndTimestampMs:   21,
+						},
+					},
+				},
+			},
+			{
+				isNaN: true,
+				rrq: prompb.ReadRequest{
+					Queries: []*prompb.Query{
+						{
+							Matchers: []*prompb.LabelMatcher{
+								{
+									Type:  prompb.LabelMatcher_EQ,
+									Name:  metricNameLabelName,
+									Value: metricName,
+								},
+							},
+							StartTimestampMs: 29,
+							EndTimestampMs:   31,
+						},
+					},
+				},
+			},
+		}
+
+		for _, c := range query {
+			r := NewPgxReader(db)
+			resp, err := r.Read(&c.rrq)
+			startMs := c.rrq.Queries[0].StartTimestampMs
+			endMs := c.rrq.Queries[0].EndTimestampMs
+			timeClause := fmt.Sprintf("time >= 'epoch'::timestamptz +interval '%d ms' AND "+
+				"time <= 'epoch'::timestamptz +interval '%d ms'", startMs, endMs)
+
+			if err != nil {
+				t.Fatalf("unexpected error while ingesting test dataset: %s", err)
+			}
+			answer := getSingleSampleValue(t, resp)
+			isStaleNaN := getBooleanSQLResult(t, db,
+				fmt.Sprintf(
+					`SELECT prom.is_stale_marker(value)
+					 FROM prom."StaleMetric"
+					 WHERE %s
```

Comment:
nit: consider using variable interpolation if it works, for sql best practice ` time >= 'epoch'::timestamptz +interval '$1 ms' AND "+
 				"time <= 'epoch'::timestamptz +interval $2 ms' `